### DISCUSSION TITLE
How does the HISTOGRAM function work?

### DISCUSSION THREAD

**Discussion by David-Salazar** [duckdb/duckdb#3684] [Q&A]
Hi, y'all!

I am a new user and have been loving DuckDB. I have one question and the docs have not been able to answer it: how does exactly the HISTOGRAM function work? Meaning, how does it choose the number of bins? Can we determine the number of bins ourselves?

Thanks!

**Comment by FrancoisLepoutre** [comment]
_Can we determine the number of bins ourselves?_

For sure, Thanks for this interesting issue and question. Histograms can certainly be organized on arbitrary scales. Having specialized tooling is quite welcome! But the way to apply them to dates is an interesting issues. Googling on "histogram()" "sql" I found nothing both formal stuff except for this:

https://docs.influxdata.com/flux/v0.x/stdlib/universe/histogram/

Which did not look convincingly generic really is possibly not what our lovely duck HISTOGRAM functions expects. I suppose that the HISTOGRAM keyword originates from a relatively recent player that in the field of big data sql or nonsql data engine that I have be in touch with:-)

Additional pointers on the way to handle the beast on arbitrary binning would be welcome!

**Comment by David-Salazar** [comment]
Thank you for responding, @FrancoisLepoutre. I am guessing that I am aiming to have some of the flexibility that [the numpy histogram function has. ](https://numpy.org/doc/stable/reference/generated/numpy.histogram.html)

**Reply by FrancoisLepoutre** [reply]
Hi David. Thanks for this great pointer. I am since a bit green on numpy. But this looks like what I'd be expecting as well when it comes to this functionality. 

Looking forward for the current official answer without peeping into the source code or the testing material. Duckdb is wonderful especially in the way it interacts with numpy and arrow by the way. As a sql dev, our great "duck" has been my entry points in both of them. But I understand it works both ways:-)

**Comment by FrancoisLepoutre** [comment]
It looks like the answer is pretty simple... Since the syntax is quite straightforward. 

I should have run a test instead for searching for an existing SQL functionality. Takes less time to open the console and run the test than googling for a generic answer.  Duckdb HISTOGRAM takes an arbitrary sql expression - I just tested plain ones, I did NOT test subqueries - and delivers an histogram as a MAP.
        Candidate functions:
        histogram(USMALLINT) -> MAP
        histogram(UINTEGER) -> MAP
        histogram(UBIGINT) -> MAP
        histogram(SMALLINT) -> MAP
        histogram(INTEGER) -> MAP
        histogram(BIGINT) -> MAP
        histogram(FLOAT) -> MAP
        histogram(DOUBLE) -> MAP
        histogram(VARCHAR) -> MAP
        histogram(TIMESTAMP) -> MAP
        histogram(TIMESTAMP WITH TIME ZONE) -> MAP
        histogram(TIMESTAMP_S) -> MAP
        histogram(TIMESTAMP_MS) -> MAP
        histogram(TIMESTAMP_NS) -> MAP

Looks pretty simple and workable. Binning buckets are not under control. Each distinct numeric delivers a single MAP entry. Of course, delivering complex binnings such as the ones possibly delivered by numpy would be a great plus. But you can anyway subcontract the functionality to numpy in case of specific needs and feed the result back into the engine in a snap. This fast bidirectional feeding link between duckdb and most common data matrix engines à la numpy, pandas, and ,of course, arrow is anyway what sets it apart! 

Some sort of high-end binning functionality for numerical information and timestamps would be a great plus. But that HISTOGRAM keyword is already quite useful as is. The SQL grammar is already quite rich. Thanks the duckdb team!

**Comment by mijalapenos** [comment]
Current implementation is very well described here:
https://cfahlgren1-sql-snippets.hf.space/histogram

It refers to the recent PR here:
https://github.com/duckdb/duckdb/pull/12590

Hope that helps anybody looking for more information.
